### PR TITLE
add the function to get and set attributes of remote models

### PR DESCRIPTION
**Pull request by YuechengLiu**
Add the function to get and set attributes of remote models.
 
i.e. we can operate the remote object in the following way,
```python
import parl
@parl.remote_class
class Agent(object):
    def __init__(self):
        self.b = 0

parl.connect('localhost:8400')
agent = Agent()
print(agent.b)  # 0
agent.b = 5
print(agent.b) # 5
```

### CODE REVIEW

File: parl/remote/remote_decorator.py [hunk 1]
```diff
@@ -179,8 +179,64 @@ def request_cpu_resource(self, global_client, max_memory):
                     cnt -= 1
                 return None
 
+            def __setattr__(self, attr, value):
+                if attr not in cls().__dict__:
```
Comment (zenghsh3) [inline parl/remote/remote_decorator.py]: We should not initialize an object here, please put this piece of logic in the remote object.

File: parl/remote/tests/get_set_attribute_test.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+from parl.utils import logger
+from parl.remote.master import Master
+from parl.remote.worker import Worker
+import time
+import threading
+import random
+
+
+@parl.remote_class
+class Actor(object):
```
Comment (zenghsh3) [inline parl/remote/tests/get_set_attribute_test.py:28]: add unittest cases:
1. `Actor` has the same attribute name and function name. try to get the attribute and call the function.
2. try to get a non-existing attribute whose name is the same as the function name.
3. try to call a non-existing function whose name is the same as the attribute names.

File: parl/remote/worker.py [hunk 3]
```diff
@@ -211,6 +210,7 @@ def _init_jobs(self, job_num):
         for _ in range(job_num):
             subprocess.Popen(command, stdout=FNULL, stderr=subprocess.STDOUT)
         FNULL.close()
+        # subprocess.Popen(command, stdout=subprocess.STDOUT, stderr=subprocess.STDOUT)
```
Comment (zenghsh3) [inline parl/remote/worker.py]: remove this line.

File: parl/remote/job.py [hunk 4]
```diff
@@ -397,24 +397,60 @@ def single_task(self, obj, reply_socket, job_address):
...
-
             tag = message[0]
-
-            if tag == remote_constants.CALL_TAG:
+            if tag in [
+                    remote_constants.CALL_TAG, remote_constants.GET_ATTRIBUTE,
+                    remote_constants.SET_ATTRIBUTE,
+                    remote_constants.CHECK_ATTRIBUTE
+            ]:
+                # if tag == remote_constants.CALL_TAG:
```
Comment (zenghsh3) [inline parl/remote/job.py]: remove this line

**Review by zenghsh3 — CHANGES_REQUESTED** [review]

File: parl/remote/remote_decorator.py [hunk 5]
```diff
@@ -228,8 +297,27 @@ def wrapper(*args, **kwargs):
 
                 return wrapper
 
+        def proxy_wrapper_func(remote_wrapper):
```
Comment (zenghsh3) [inline parl/remote/remote_decorator.py:269]: Add some comments to explain why we need this proxy wrapper.

**Review by zenghsh3 — CHANGES_REQUESTED** [review]

File: parl/remote/remote_decorator.py (hunk 1 above)
Comment (YuechengLiu) [reply parl/remote/remote_decorator.py]: fixed

File: parl/remote/tests/get_set_attribute_test.py (hunk 2 above)
Comment (YuechengLiu) [reply parl/remote/tests/get_set_attribute_test.py:28]: added

File: parl/remote/remote_decorator.py [hunk 6]
```diff
@@ -179,8 +179,77 @@ def request_cpu_resource(self, global_client, max_memory):
...
+                is_attribute = self.check_attribute(attr)
+                if is_attribute:
+                    self.internal_lock.acquire()
+                    self.job_socket.send_multipart(
+                        [remote_constants.GET_ATTRIBUTE,
+                         to_byte(attr)])
+                    message = self.job_socket.recv_multipart()
+                    tag = message[0]
+
+                    if tag == remote_constants.NORMAL_TAG:
```
Comment (zenghsh3) [inline parl/remote/remote_decorator.py]: line 224-252 is repeated to line 266-296, can we provide a common function.

Comment (YuechengLiu) [reply parl/remote/remote_decorator.py]: modified

**Review by zenghsh3 — APPROVED** [review]
LGTM

**Pull request merged**